File: app/src/main/java/com/codylimber/fieldphenology/ui/components/StatusBadge.kt
package com.codylimber.fieldphenology.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import com.codylimber.fieldphenology.data.model.SpeciesStatus
import com.codylimber.fieldphenology.ui.theme.*

/**
 * The one chip primitive the app's badges are built from: a pill with a tonal
 * background and matching text color, so it stays legible in both themes
 * without the white-on-pastel washout that solid-fill badges produce in light
 * mode.
 */
@Composable
private fun ToneChip(
    label: String,
    role: ToneRole,
    modifier: Modifier = Modifier,
) {
    val dark = LocalIsDarkTheme.current
    Text(
        text = label,
        color = role.onContainer(dark),
        style = MaterialTheme.typography.labelSmall,
        modifier = modifier
            .clip(PillShape)
            .background(role.container(dark))
            .padding(horizontal = 8.dp, vertical = 3.dp)
    )
}

@Composable
private fun statusRole(status: SpeciesStatus): Pair<String, ToneRole> = when (status) {
    SpeciesStatus.PEAK -> "Peak" to StatusPeakRole
    SpeciesStatus.ACTIVE -> "Active" to StatusActiveRole
    SpeciesStatus.EARLY -> "Early" to StatusEdgeRole
    SpeciesStatus.LATE -> "Late" to StatusEdgeRole
    SpeciesStatus.INACTIVE -> "Inactive" to StatusInactiveRole
}

@Composable
fun StatusBadge(status: SpeciesStatus, modifier: Modifier = Modifier) {
    val (label, role) = statusRole(status)
    ToneChip(label = label, role = role, modifier = modifier)
}

private fun activityRole(percent: Int): ToneRole = when {
    percent >= 80 -> StatusPeakRole
    percent >= 20 -> StatusActiveRole
    percent > 0 -> StatusEdgeRole
    else -> StatusInactiveRole
}

@Composable
fun ActivityBadge(percent: Int, modifier: Modifier = Modifier) {
    ToneChip(label = "$percent%", role = activityRole(percent), modifier = modifier)
}

@Composable
fun ActivityDot(percent: Int, modifier: Modifier = Modifier) {
    val dark = LocalIsDarkTheme.current
    val color = activityRole(percent).solid(dark)
    // Fill tracks abundance: solid when the species is well into its season,
    // hollow when it is only just on or off.
    Box(
        modifier = modifier
            .size(10.dp)
            .clip(CircleShape)
            .background(if (percent >= 50) color else color.copy(alpha = 0.28f))
            .padding(1.5.dp)
    ) {
        if (percent in 1..49) {
            Box(
                Modifier
                    .size(5.dp)
                    .clip(CircleShape)
                    .background(color)
            )
        }
    }
}

private fun rarityRole(rarity: String): ToneRole? = when (rarity) {
    "Common" -> RarityCommonRole
    "Uncommon" -> RarityUncommonRole
    "Rare" -> RarityRareRole
    else -> null
}

@Composable
fun RarityDot(rarity: String, modifier: Modifier = Modifier) {
    val dark = LocalIsDarkTheme.current
    val color = rarityRole(rarity)?.solid(dark) ?: MaterialTheme.colorScheme.onSurfaceVariant
    Box(
        modifier = modifier
            .size(7.dp)
            .clip(CircleShape)
            .background(color)
    )
}

@Composable
fun RarityChip(rarity: String, modifier: Modifier = Modifier) {
    val role = rarityRole(rarity) ?: return
    ToneChip(label = rarity, role = role, modifier = modifier)
}

@Composable
fun ConservationBadge(status: String?, modifier: Modifier = Modifier) {
    if (status == null) return
    val code = status.uppercase()
    val dark = LocalIsDarkTheme.current
    // Only the IUCN codes get a badge; anything else is dataset noise.
    val solid = when (code) {
        "LC" -> ConservationLC
        "NT" -> ConservationNT
        "VU" -> ConservationVU
        "EN" -> ConservationEN
        "CR", "EW", "EX" -> ConservationCR
        else -> return
    }
    // Conservation codes are rare enough on screen that they read better as an
    // outlined chip than as another filled one competing with the status badge.
    Text(
        text = code,
        color = if (dark) solid else solid.darken(),
        style = MaterialTheme.typography.labelSmall,
        modifier = modifier
            .clip(PillShape)
            .background(solid.copy(alpha = if (dark) 0.16f else 0.20f))
            .padding(horizontal = 8.dp, vertical = 3.dp)
    )
}

/** Pull a bright dark-theme accent down to something readable on white. */
private fun Color.darken(factor: Float = 0.62f) =
    Color(red * factor, green * factor, blue * factor, alpha)

File: ios/Manakin/UI/Components/StatusBadge.swift
import SwiftUI

struct StatusBadge: View {
    let status: SpeciesStatus

    private var label: String {
        switch status {
        case .peak: return "Peak"
        case .active: return "Active"
        case .early: return "Early"
        case .late: return "Late"
        case .inactive: return "Inactive"
        }
    }

    private var color: Color {
        switch status {
        case .peak: return .statusPeak
        case .active: return .statusActive
        case .early, .late: return .statusEarlyLate
        case .inactive: return .statusInactive
        }
    }

    var body: some View {
        Text(label)
            .font(.system(size: 11, weight: .semibold))
            .foregroundColor(.white)
            .padding(.horizontal, 6)
            .padding(.vertical, 2)
            .background(color.opacity(0.85))
            .clipShape(RoundedRectangle(cornerRadius: 4))
    }
}

struct ActivityBadge: View {
    let percent: Int

    private var color: Color {
        switch percent {
        case 80...: return .statusPeak
        case 20...: return .statusActive
        case 1...: return .statusEarlyLate
        default: return .statusInactive
        }
    }

    var body: some View {
        Text("\(percent)%")
            .font(.system(size: 11, weight: .semibold))
            .foregroundColor(.white)
            .padding(.horizontal, 6)
            .padding(.vertical, 2)
            .background(color.opacity(0.85))
            .clipShape(RoundedRectangle(cornerRadius: 4))
    }
}

struct ActivityDot: View {
    let percent: Int

    private var color: Color {
        switch percent {
        case 80...: return .statusPeak
        case 20...: return .statusActive
        case 1...: return .statusEarlyLate
        default: return .statusInactive
        }
    }

    private var symbol: String {
        switch percent {
        case 50...: return "\u{25CF}"   // filled circle
        case 1...: return "\u{25D0}"    // half circle
        default: return "\u{25CB}"      // empty circle
        }
    }

    var body: some View {
        Text(symbol)
            .font(.system(size: 14))
            .foregroundColor(color)
    }
}

struct RarityDot: View {
    let rarity: String
    @Environment(\.appColors) private var colors

    private var color: Color {
        switch rarity {
        case "Common": return .rarityCommon
        case "Uncommon": return .rarityUncommon
        case "Rare": return .rarityRare
        default: return colors.onSurfaceVariant
        }
    }

    var body: some View {
        Text("\u{25CF}")
            .font(.system(size: 12))
            .foregroundColor(color)
    }
}

struct RarityChip: View {
    let rarity: String

    private var color: Color? {
        switch rarity {
        case "Common": return .rarityCommon
        case "Uncommon": return .rarityUncommon
        case "Rare": return .rarityRare
        default: return nil
        }
    }

    var body: some View {
        if let color {
            Text(rarity)
                .font(.system(size: 10, weight: .semibold))
                .foregroundColor(.white)
                .padding(.horizontal, 5)
                .padding(.vertical, 1)
                .background(color.opacity(0.75))
                .clipShape(RoundedRectangle(cornerRadius: 4))
        }
    }
}

struct ConservationBadge: View {
    let status: String?

    private static let iucnStatuses: Set<String> = ["LC", "NT", "VU", "EN", "CR", "EW", "EX"]

    private var color: Color? {
        guard let status = status?.uppercased(),
              Self.iucnStatuses.contains(status) else { return nil }
        switch status {
        case "LC": return .conservationLC
        case "NT": return .conservationNT
        case "VU": return .conservationVU
        case "EN": return .conservationEN
        case "CR", "EW", "EX": return .conservationCR
        default: return nil
        }
    }

    var body: some View {
        if let status = status?.uppercased(),
           Self.iucnStatuses.contains(status),
           let color {
            Text(status)
                .font(.system(size: 10, weight: .bold))
                .foregroundColor(.white)
                .padding(.horizontal, 5)
                .padding(.vertical, 1)
                .background(color.opacity(0.85))
                .clipShape(RoundedRectangle(cornerRadius: 4))
        }
    }
}
